### ISSUE TITLE
When I run the GRPO demo, I find that format_reward is always 0！！！

### ISSUE THREAD

**Issue by asirgogogo** [huggingface/open-r1#235]
```shell
ACCELERATE_LOG_LEVEL=info accelerate launch --config_file recipes/accelerate_configs/zero2.yaml \
    --num_processes=7 src/open_r1/grpo.py \
    --config recipes/Qwen2.5-Math-7B/grpo/config_simple_rl.yaml
```

I'm sure that I didn't modify the parameters in “config_simple_rl.yaml”.

<img width="1577" alt="Image" src="https://github.com/user-attachments/assets/a2dea1cd-43a9-4edb-8e74-c9676981eb4b" />

**Comment by tpoisonooo** [comment]
format_reward is matching exact “<think>.. <answer>” format, return 0 or 1.

I guess 
- this reward_func curve is too sharp
- the training dataset (NuminaMath-TIR)  does not have this response format.

maybe a good PR ?

**Comment by tpoisonooo** [comment]
```python
def format_reward(completions, **kwargs):
    """Reward function that checks if the completion has a specific format."""
    pattern = r"^<think>.*?</think><answer>.*?</answer>$"
    completion_contents = [completion[0]["content"] for completion in completions]
    matches = [re.match(pattern, content) for content in completion_contents]
    return [1.0 if match else 0.0 for match in matches]
```

**Comment by ytchx1999** [comment]
same

![Image](https://github.com/user-attachments/assets/ce87dee2-02c4-49d9-985a-839017c9277c)

**Comment by hellen9527** [comment]
I want to ask how big a machine is needed to train a 7B model. I have 8 L20 GPUs, and when I train for 10 steps, the loss is 0, and the reward is also 0, though the accuracy_reward is not 0. I’m not sure why. However, at step 15, I get an error saying that the GPU memory is insufficient. Could you let me know what machine you're using?

**Comment by asirgogogo** [comment]
@hellen9527 I use 8xA800 GPUs

**Comment by asirgogogo** [comment]
@tpoisonooo According to the settings, “DigitalLearningGmbH/MATH-lighteval” dataset is used.

**Comment by hellen9527** [comment]
可能是有换行符导致正则表达式判断为错了，返回0分，另外一个问题里说用这个格式分换下面的正则表达式效果更好
`
pattern = r"<think>.*?</think>\s*<answer>.*?</answer>"
completion_contents = [completion[0]["content"] for completion in completions]
matches = [re.match(pattern, content, re.DOTALL) for content in completion_contents]
`

**Comment by tpoisonooo** [comment]
> 可能是有换行符导致正则表达式判断为错了，返回0分，另外一个问题里说用这个格式分换下面的正则表达式效果更好
> 
> ```
> pattern = r"<think>.*?</think>\s*<answer>.*?</answer>"
> completion_contents = [completion[0]["content"] for completion in completions]
> matches = [re.match(pattern, content, re.DOTALL) for content in completion_contents]
> ```

下面的 expr 没卡那么严，确实能缓解。

**Comment by brishtiteveja** [comment]
Related node: Please also check whether you are using cuda 12.4 (as it is shown as Caution in ReadMe). I was initially using cuda 12.1 and then switched to 12.4

**Comment by izlley** [comment]
The issue has been fixed. https://github.com/huggingface/open-r1/pull/238

**Comment by Moyhub** [comment]
@izlley After I modified it by pressing #238, the forwar reward was still almost zero when I used qwen2.5-math-7B for GRPO training, as follows:

![Image](https://github.com/user-attachments/assets/f23978df-45e6-4c1e-b46e-dee9959ae5e7)

**Comment by ZhenweiAn** [comment]
Perhaps there should have a cold start for fomat before training grpo

**Comment by yuleiqin** [comment]
same here, with the new format reward, the model still fails to deliever correct formats

`def format_reward(completions, **kwargs):
    """Reward function that checks if the completion has a specific format."""
    pattern = r"^<think>.*?</think>\s*<answer>.*?</answer>$"
    completion_contents = [completion[0]["content"] for completion in completions]
    matches = [re.match(pattern, content, re.DOTALL | re.MULTILINE) for content in completion_contents]
    return [1.0 if match else 0.0 for match in matches]`

<img width="1655" alt="Image" src="https://github.com/user-attachments/assets/e8428138-e6e9-464b-8bbd-5d9579dc18ab" />

8 H20 GPUs, DeepSeek-R1-Distill-Qwen-7B-GRPO

> [@izlley](https://github.com/izlley) After I modified it by pressing [#238](https://github.com/huggingface/open-r1/pull/238), the forwar reward was still almost zero when I used qwen2.5-math-7B for GRPO training, as follows:
> 
> ![Image](https://github.com/user-attachments/assets/f23978df-45e6-4c1e-b46e-dee9959ae5e7)

**Comment by yihao-123** [comment]
I use the Qwen2.5-7B-Base， and it can output the correct format. However, the Qwen2.5-7B-Math-Instruct cannot output the correct format.

**Comment by peter-peng-w** [comment]
> I use the Qwen2.5-7B-Base， and it can output the correct format. However, the Qwen2.5-7B-Math-Instruct cannot output the correct format.

Interesting, my format reward barely improve when training on Qwen2.5-7b-base. Did you modify any part of the training arguments in the recipes?

**Comment by yihao-123** [comment]
> > I use the Qwen2.5-7B-Base， and it can output the correct format. However, the Qwen2.5-7B-Math-Instruct cannot output the correct format.
> 
> Interesting, my format reward barely improve when training on Qwen2.5-7b-base. Did you modify any part of the training arguments in the recipes?

I didn't modify any argument. You can try to inference one case using the system prompt in the grpo.py.

**Comment by DogeWatch** [comment]
Have you been monitoring your completion_length? In my experiments, I've noticed that it consistently remains at the upper limit I set after a certain number of training steps. Could this be the reason why it doesn't meet the reward requirements in terms of format?

<img width="354" alt="Image" src="https://github.com/user-attachments/assets/83e9df54-d4e3-4584-b1f0-3071d14a2b0b" />

**Comment by TianpengBu** [comment]
Check the system promot in recipes/Qwen2.5-Math-7B/grpo/config_simple_rl.yaml. The prompt suggests the model to give output without <answer> tag, but the reward_func augment selects both accuracy and format. You can do the following to solve this: 

1. change the system prompt and make it give both <think> and <answer> tags. 
2. keep this system prompt and disable the format reward.

**Comment by TianpengBu** [comment]
Check the system prompt in recipes/Qwen2.5-Math-7B/grpo/config_simple_rl.yaml. The prompt suggests the model to give output without tag, but the reward_func augment selects both accuracy and format. You can do the following to solve this:

change the system prompt and make it give both and tags.
keep this system prompt and disable the format reward.

**Comment by TianpengBu** [comment]
Check the system prompt in recipes/Qwen2.5-Math-7B/grpo/config_simple_rl.yaml. The prompt suggests the model to give output without the <answer> tag, but the reward_func augment selects both accuracy and format rewards. You can do the following to solve this:

change the system prompt and make it give both <think> and <answer> tags.
keep this system prompt and disable the format reward.

**Comment by cainiaoup** [comment]
> I want to ask how big a machine is needed to train a 7B model. I have 8 L20 GPUs, and when I train for 10 steps, the loss is 0, and the format_reward is also 0, though the accuracy_reward is not 0. I’m not sure why. However, at step 15, I get an error saying that the GPU memory is insufficient. Could you let me know what machine you're using?

8*48G(L40s) + zeros3 + batch size 1 can work for llama-8b. context length 8k + 4K
8*800G(H100) + zeros3 + batch size 4 can work for  llama-8b. context length 8k + 4K

**Comment by KyrieXDL** [comment]
It may be related to the base model and system prompt. Utilizing the Qwen2.5-7b-base, the format reward will gradually increase. But the accuracy reward is relatively slow.

![Image](https://github.com/user-attachments/assets/bd64e8e8-8db5-4997-a0e0-ea9c20d62f1f)

**Comment by SiqingHe** [comment]
> > I want to ask how big a machine is needed to train a 7B model. I have 8 L20 GPUs, and when I train for 10 steps, the loss is 0, and the format_reward is also 0, though the accuracy_reward is not 0. I’m not sure why. However, at step 15, I get an error saying that the GPU memory is insufficient. Could you let me know what machine you're using?
> 
> 8_48G(L40s) + zeros3 + batch size 1 can work for llama-8b. context length 8k + 4K 8_800G(H100) + zeros3 + batch size 4 can work for llama-8b. context length 8k + 4K

is batch size 1 works when training ?

**Comment by PETEPEtrek** [comment]
> same here, with the new format reward, the model still fails to deliever correct formats
> 
> `def format_reward(completions, **kwargs): """Reward function that checks if the completion has a specific format.""" pattern = r"^<think>.*?</think>\s*<answer>.*?</answer>$" completion_contents = [completion[0]["content"] for completion in completions] matches = [re.match(pattern, content, re.DOTALL | re.MULTILINE) for content in completion_contents] return [1.0 if match else 0.0 for match in matches]`
> 
> <img alt="Image" width="1655" src="https://private-user-images.githubusercontent.com/110647771/411995240-e8428138-e6e9-464b-8bbd-5d9579dc18ab.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY2MTQ3NDEsIm5iZiI6MTc0NjYxNDQ0MSwicGF0aCI6Ii8xMTA2NDc3NzEvNDExOTk1MjQwLWU4NDI4MTM4LWU2ZTktNDY0Yi04YmJkLTVkOTU3OWRjMThhYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwN1QxMDQwNDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMmFmNGQ0NDVlNGE1OWJmMDAzZTk1MzAwMDc4NjIyZDY2M2I2NDE3NjkzOWVkZTE2N2Q3MmU4MTdlMDMxNDZhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2rnvKKJRiTyTWlQlxBwgp549H7nl1x45lNZ6bBOYVy8">
> 8 H20 GPUs, DeepSeek-R1-Distill-Qwen-7B-GRPO
> 
> > [@izlley](https://github.com/izlley) After I modified it by pressing [#238](https://github.com/huggingface/open-r1/pull/238), the forwar reward was still almost zero when I used qwen2.5-math-7B for GRPO training, as follows:
> > ![Image](https://github.com/user-attachments/assets/f23978df-45e6-4c1e-b46e-dee9959ae5e7)

Hi, did you manage to solve this problem? My format reward is like 

pattern = r"^<think>.*?</think>\s*<answer>.*?</answer>$"

    # Extract the content from each completion
    completion_contents = [completion[0]['content'] for completion in completions]

    # Check if each completion matches the pattern
    matches = [re.match(pattern, content, re.DOTALL | re.MULTILINE) for content in completion_contents]

    # Reward 1.0 for correct format, 0.0 otherwise
    format_rewards = [1.0 if match else 0.0 for match in matches]


But i still have 0 format reward

**Comment by lewtun** [comment]
Hello everyone, we are working on a better recipe for R1-Zero-like training in this PR: https://github.com/huggingface/open-r1/pull/569

The main lesson is that (a) one should use a better dataset like DAPO's and (b) mask overlong completions. An example config that produces stable training and good evals is this one: https://github.com/huggingface/open-r1/pull/569/files#diff-a3839d533196dac34b500ef5a031d736f2f088d2c8996ddfded442fc528ef77c
